staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/matchconditions/matcher.go

```diff
@@ -0,0 +1,144 @@
+limitations under the License.
+*/
+
+package matchconditions
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/google/cel-go/cel"
+	celtypes "github.com/google/cel-go/common/types"
+
+	v1 "k8s.io/api/admissionregistration/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apiserver/pkg/admission"
+	admissionmetrics "k8s.io/apiserver/pkg/admission/metrics"
+	celplugin "k8s.io/apiserver/pkg/admission/plugin/cel"
+	celconfig "k8s.io/apiserver/pkg/apis/cel"
+	"k8s.io/apiserver/pkg/authorization/authorizer"
+	"k8s.io/klog/v2"
+)
+
+var _ celplugin.ExpressionAccessor = &MatchCondition{}
+
+// MatchCondition contains the inputs needed to compile, evaluate and match a cel expression
+type MatchCondition v1.MatchCondition
+
+func (v *MatchCondition) GetExpression() string {
+	return v.Expression
+}
+
+func (v *MatchCondition) ReturnTypes() []*cel.Type {
+	return []*cel.Type{cel.BoolType}
+}
+
+var _ Matcher = &matcher{}
+
+// matcher evaluates compiled cel expressions and determines if they match the given request or not
+type matcher struct {
+	filter      celplugin.Filter
+	authorizer  authorizer.Authorizer
+	failPolicy  *v1.FailurePolicyType
+	matcherType string
+	objectName  string
+}
+
+func NewMatcher(filter celplugin.Filter, authorizer authorizer.Authorizer, failPolicy *v1.FailurePolicyType, matcherType, objectName string) Matcher {
+	return &matcher{
+		filter:      filter,
+		authorizer:  authorizer,
+		failPolicy:  failPolicy,
+		matcherType: matcherType,
+		objectName:  objectName,
+	}
+}
+
+func (m *matcher) Match(ctx context.Context, versionedAttr *admission.VersionedAttributes, versionedParams runtime.Object) MatchResult {
+	var f v1.FailurePolicyType
+	if m.failPolicy == nil {
+		f = v1.Fail
+	} else {
+		f = *m.failPolicy
+	}
+
+	evalResults, err := m.filter.ForInput(ctx, versionedAttr, celplugin.CreateAdmissionRequest(versionedAttr.Attributes), celplugin.OptionalVariableBindings{
+		VersionedParams: versionedParams,
+		Authorizer:      m.authorizer,
+	}, celconfig.RuntimeCELCostBudgetMatchConditions)
+
+	//TODO: add event to webhook object for error
+	if err != nil {
+		// filter returning error is unexpected and not an evaluation error so not incrementing metric here
+		if f == v1.Fail {
+			return MatchResult{
+				Error: err,
+			}
+		} else {
+			return MatchResult{
+				Matches: true,
+			}
+		}
+	}
+
+	errorList := []error{}
+	for _, evalResult := range evalResults {
+		matchCondition, ok := evalResult.ExpressionAccessor.(*MatchCondition)
+		if !ok {
+			klog.Error("Invalid type conversion to MatchCondition")
+			continue
+		}
+		if evalResult.Error != nil {
+			errorList = append(errorList, evalResult.Error)
+			//TODO: what's the best way to handle this metric since its reused by VAP for match conditions
+			admissionmetrics.Metrics.ObserveMatchConditionEvalError(ctx, m.objectName, m.matcherType)
+		}
+		if evalResult.EvalResult == celtypes.False {
```

Comment:
So the CEL filter evaluates all of the compilationResults at once and returns a list of evaluationResults. This component could be reworked to do one at a time but it would be difficult to do now as its reused by VAP for validation and not just the matcher. Maybe we take a TODO for that performance enhancement? VAP may also be able to benefit from this 

cc: @jpbetz @cici37

Reply:
yeah, that's fine as a follow-up, add it to the list to look at for beta